src/PostgREST/Request/ApiRequest.hs

```diff
@@ -390,21 +392,14 @@ userApiRequest conf@AppConfig{..} dbStructure req reqBody
         where
           l = fromMaybe 0 $ rangeLimit rl
           o = rangeOffset ro
-  ranges = M.insert "limit" (rangeIntersection headerRange (fromMaybe allRange (M.lookup "limit" urlRange))) urlRange
-
-  limitFromQparams :: [(Text, Maybe ByteString)] -> [Maybe Integer]
-  limitFromQparams [] = []
-  limitFromQparams ((k, v):qs) = if isLimitQueryParameter then (readMaybe . toS =<< v):remainingLimits else remainingLimits
-    where isLimitQueryParameter = isJust v && endingIn ["limit"] k
-          remainingLimits = limitFromQparams qs
-
-  invalidLimit :: Bool
-  invalidLimit = foldl trueIfNegative False limits
-    where limits = limitFromQparams qParams
-          trueIfNegative = \ acc lim
-                              -> case lim of
-                                  Just n  -> (n < 0) || acc
-                                  Nothing -> acc
+  limitRange = fromMaybe allRange (M.lookup "limit" urlRange)
+  headerAndLimitRange = rangeIntersection headerRange limitRange
+  -- If it has limit=0 in the params, bypass all ranges and send
+  -- the limit zero range (0 <= x <= -1)
+  ranges = M.insert "limit" (if hasLimitZero limitRange then limitZeroRange else headerAndLimitRange) urlRange
+
+  -- The only emptyRange allowed is the limit zero range (limit=0 in the params)
+  isInvalidRange = topLevelRange == emptyRange && not (hasLimitZero limitRange)
```

Comment:
A comment mentioning that `limit=0` is only valid through query params - and not through the Range header - would be good.